modules/dns/dns_api.rb

```diff
@@ -5,36 +5,8 @@ class Api < ::Sinatra::Base
     authorize_with_ssl_client
 
     def dns_setup(opts)
-      raise "Smart Proxy is not configured to support DNS" unless Proxy::Dns::Plugin.settings.enabled
-      case Proxy::Dns::Plugin.settings.dns_provider
-      when "dnscmd"
-        require 'dns/providers/dnscmd'
-        @server = Proxy::Dns::Dnscmd.new(opts.merge(
-          :server => Proxy::Dns::Plugin.settings.dns_server,
-          :ttl => Proxy::Dns::Plugin.settings.dns_ttl
-        ))
-      when "nsupdate"
-        require 'dns/providers/nsupdate'
-        @server = Proxy::Dns::Nsupdate.new(opts.merge(
-          :server => Proxy::Dns::Plugin.settings.dns_server,
-          :ttl => Proxy::Dns::Plugin.settings.dns_ttl
-        ))
-      when "nsupdate_gss"
-        require 'dns/providers/nsupdate_gss'
-        @server = Proxy::Dns::NsupdateGSS.new(opts.merge(
-          :server => Proxy::Dns::Plugin.settings.dns_server,
-          :ttl => Proxy::Dns::Plugin.settings.dns_ttl,
-          :tsig_keytab => Proxy::Dns::Plugin.settings.dns_tsig_keytab,
-          :tsig_principal => Proxy::Dns::Plugin.settings.dns_tsig_principal
-        ))
-      when "virsh"
-        require 'dns/providers/virsh'
-        @server = Proxy::Dns::Virsh.new(opts.merge(
-          :virsh_network => Proxy::SETTINGS.virsh_network
-        ))
-      else
-        log_halt 400, "Unrecognized or missing DNS provider: #{Proxy::Dns::Plugin.settings.dns_provider || "MISSING"}"
-      end
+      factory = ::Proxy::Plugins.find_provider_factory(Proxy::Dns::Plugin.settings.dns_provider)
+      @server = factory.call(opts)
```

Comment:
Possibly, but I think it's better to delegate all responsibility for creation/returning of dns providers (in this case) to the factory itself, as only it knows all the nitty-gritty involved, including potential side effects and such.